Source/WebKit/Shared/JavaScriptEvaluationResult.cpp

```diff
@@ -55,15 +55,30 @@ class JavaScriptEvaluationResult::JSExtractor {
     Map takeMap() { return WTFMove(m_map); }
     std::optional<JSObjectID> addObjectToMap(JSGlobalContextRef, JSValueRef);
 
-    bool isSafeToRecurse();
-
 private:
-    std::optional<Value> toValue(JSGlobalContextRef, JSValueRef);
+    void extractJSValue(JSGlobalContextRef, JSValueRef);
+
+    struct UnsupportedTypeTag { };
+    struct PendingCollectionTag { };
+    using ExtractionResult = Variant <
+        Value,
+        UnsupportedTypeTag,
+        PendingCollectionTag
+    >;
+
+    ExtractionResult jsValueToExtractedValue(JSGlobalContextRef, JSValueRef);
+    void addJSValueToExtractLater(JSGlobalContextRef, JSValueRef);
 
     Map m_map;
-    HashMap<Protected<JSValueRef>, JSObjectID> m_objectsInMap;
-    StackCheck m_stackCheck;
-    bool m_didStackOverflow { false };
+    HashMap<Protected<JSValueRef>, JSObjectID> m_valuesInMap;
+
+    static constexpr size_t maximumNestingLevel { 40000 };
+    size_t m_currentNestingLevel { 0 };
+
+    HashSet<Protected<JSValueRef>> m_futureJSValuesToExtract;
+    HashSet<Protected<JSValueRef>> m_currentJSValuesToExtract;
```

Comment:
I'm pretty sure you can combine these into a single `Vector<Protected<JSValueRef>>`. I believe they are only hash tables for the sake of the uniqueness check, and if you agree that the uniqueness check can be removed (and instead done later, when we pop a value off the stack I am proposing), then this can become simpler and more efficient.